AWS database access configurator

gabrielcorado opened the pull request:
Database configure bootstrap implementation from [RFD46](https://github.com/gravitational/teleport/blob/master/rfd/0046-database-access-config.md).

Some notes for reviewers:
* The interaction with AWS, data validation, and policy generation is implemented inside the `lib/configurators/databases` package;
* Policies were taken from [the documentation page](https://goteleport.com/docs/ver/8.0/database-access/guides/rds/) and tested;
* Documentation updates will be done in a separated PR;

gabrielcorado committed:
feat: aws database configurator

r0mant commented on lib/configurators/databases/aws.go:
```diff
@@ -0,0 +1,656 @@
+
+	// If no versions were found, we need to create a new policy.
+	if trace.IsNotFound(err) {
+		resp, err := iamClient.CreatePolicyWithContext(context.Background(), &iam.CreatePolicyInput{
```
Let's pass the context all the way from the caller. Same throughout.

r0mant commented on lib/configurators/databases/aws.go:
```diff
@@ -0,0 +1,656 @@
+		return "", nil, trace.Wrap(err)
+	}
+
+	getPolicyResp, err := client.GetPolicy(&iam.GetPolicyInput{PolicyArn: aws.String(policyArn)})
```
Let's use XXXWithContext SDK methods and pass context object everywhere from the caller.

r0mant commented on lib/configurators/databases/aws.go:
```diff
@@ -0,0 +1,656 @@
+		return "", trace.Wrap(err)
+	}
+
+	//
```
Unfinished comment?

r0mant commented on lib/configurators/databases/aws.go:
```diff
@@ -0,0 +1,656 @@
+// Since policies can hold only five versions, we need to delete a policy
+// version (if the limit is reached) and create a new version.
+// Ref: https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies_managed-versioning.html#version-limits
+func (a *awsConfigurator) upsertPolicyDocument(documentName, documentDescription string, document *awslib.PolicyDocument) (string, error) {
```
Feels like this method should not be a part of configurator but rather some helper method that accepts AWS client and other parameters or maybe some a method of AWSClient object that wraps AWS clients and provides helper methods like this one.

r0mant commented on lib/configurators/databases/aws.go:
```diff
@@ -0,0 +1,656 @@
+	}
+
+	// Check number of policy versions and delete one if necessary.
+	if len(versions) == 5 {
```
`5` should probably be a constant somewhere.

r0mant commented on lib/configurators/databases/aws.go:
```diff
@@ -0,0 +1,656 @@
+	// Check number of policy versions and delete one if necessary.
+	if len(versions) == 5 {
+		// Sort versions based on create date.
+		sort.Slice(versions, func(i, j int) bool {
```
Can this logic be factored out into `DeleteOldestPolicy` or something similar e.g. on that AWSClient wrapper I suggested above.

r0mant commented on lib/configurators/databases/aws.go:
```diff
@@ -0,0 +1,656 @@
+
+	var err error
+	a.awsSession, err = awssession.NewSessionWithOptions(awssession.Options{
+		// TODO: remove this, only used to test locally
+		// Config: aws.Config{
+		// 	Endpoint: aws.String("http://localhost:4566"),
+		// },
```
Don't forget to remove this.

r0mant commented on lib/configurators/databases/aws.go:
```diff
@@ -0,0 +1,656 @@
+
+// buildPolicyDocuments generates policy and policy boundary document based on
+// flags and fileConfig.
+func buildPolicyDocuments(targetType targetType, flags BootstrapFlags, fileConfig *config.FileConfig, identity awslib.Identity) (*awslib.PolicyDocument, *awslib.PolicyDocument, error) {
```
I think there's too much going on in this method, it's pretty hard to follow.

First, I would split it in 2: separate for generating policy document and boundary document. This will align pretty well with my suggestion above about having separate "steps".

Also, I would refactor it so the flow/logic is more clear, for example:

```go
func buildPolicyDocument(...) (*awslib.PolicyDocument, error) {
  var statements []xxx
  if rdsEnabled() {
    rdsStatements, err := getRDSStatements()
    // check err
    statements = append(statements, rdsStatements...)
  }
  if auroraEnabled() {
    // getAuroraStatements()
  }
  if redshiftEnabled() {
    // getRedshiftStatements()
  }
  return awslib.NewPolicyDocument(statements)
}

func buildBoundaryDocument(...) (*awslib.PolicyDocument, error) {
  // similar to above
}
```

r0mant commented on lib/configurators/databases/aws.go:
```diff
@@ -0,0 +1,656 @@
+
+// formatPolicyDocument formats the PolicyDocument in a "friendly" format, which
+// can be presented to end users.
+func formatPolicyDocument(policy *awslib.PolicyDocument) (string, error) {
```
Can this be a `Marshal` method on the PolicyDocument?

r0mant commented on lib/configurators/databases/aws.go:
```diff
@@ -0,0 +1,656 @@
+}
+
+// encodePolicyDocument encode PolicyDocument into JSON.
+func encodePolicyDocument(policy *awslib.PolicyDocument) (string, error) {
```
Is this needed? Can we just always marshal it with indent so we can use the same method for both presenting to users and sending via API?

r0mant commented on lib/configurators/databases/aws.go:
```diff
@@ -0,0 +1,656 @@
+}
+
+// wrapAWSError wraps an AWS error accordingly.
+func wrapAWSError(err error) error {
```
Is this method needed? Why not just call awslib.ConvertAWSError(e)?

gabrielcorado replied:
It was needed since the `ConvertAWSError` expected an `awserr.RequestFailure`. I've updated the function to receive an `error` and convert internally, so we don't need this function anymore.

r0mant commented on lib/configurators/databases/aws.go:
```diff
@@ -0,0 +1,656 @@
+}
+
+// targetType represents types that will have the policies attached to.
+type targetType int
```
Can this be a string? Then you won't need String() method.

I would also consider moving this to lib/cloud/aws/identity.go and calling it IdentityType or something.

gabrielcorado replied:
This type is no longer necessary, and the configurator now relies on the `Identity` type.

r0mant commented on lib/configurators/databases/databases.go:
```diff
@@ -0,0 +1,103 @@
+	AttachToRole string
+	// IncludeRedshiftWildcard indicates if Redshift permissions must be
+	// included using a wildcard.
+	IncludeRedshiftWildcard bool
```
I'm not sure what this flag is for?

gabrielcorado replied:
It was left after a refactoring.

r0mant commented on lib/configurators/databases/databases.go:
```diff
@@ -0,0 +1,103 @@
+	// IncludeRedshiftWildcard indicates if Redshift permissions must be
+	// included using a wildcard.
+	IncludeRedshiftWildcard bool
+	// ForceRDSPermissions forces the presence of RDS permissions.
+	ForceRDSPermissions bool
+	// ForceAuroraPermissions forces the presence of Redshift permissions.
+	ForceRedshiftPermissions bool
```
Are these needed? Can we simplify and just determine permissions based on the config file alone for now? Would probably simplify a few things.

gabrielcorado replied:
We need them to use the configurator for the `aws create-iam` and `aws print-iam` commands. Those commands define if the policy should have RDS/Redshift permissions using flags (instead of the config file).

r0mant commented on lib/configurators/databases/databases.go:
```diff
@@ -0,0 +1,103 @@
+	Err error
+}
+
+// Configurator
```
Unfinished comment.

r0mant commented on tool/teleport/common/teleport.go:
```diff
@@ -205,6 +215,35 @@ func Run(options Options) (app *kingpin.Application, executedCommand string, con
 	dbStartCmd.Flag("diag-addr", "Start diagnostic prometheus and healthz endpoint.").StringVar(&ccf.DiagnosticAddr)
 	dbStartCmd.Flag("insecure", "Insecure mode disables certificate validation").BoolVar(&ccf.InsecureMode)
 	dbStartCmd.Alias(dbUsageExamples) // We're using "alias" section to display usage examples.
+	dbConfigure := dbCmd.Command("configure", "Database agent bootstrap.")
```
```suggestion
	dbConfigure := dbCmd.Command("configure", "Bootstraps database service configuration and cloud permissions.")
```

r0mant commented on tool/teleport/common/teleport.go:
```diff
@@ -490,6 +535,255 @@ func onForward() error {
+// user confirms. The confirmation is done by typing "yes" or "y". The caller
+// can provide an autoConfirm param which will skip the user input and return
+// `true`.
+func requestConfirmation(autoConfirm bool) (bool, error) {
```
We have lib/utils/prompt/confirmation.go.

r0mant commented on tool/teleport/common/teleport.go:
```diff
@@ -490,6 +535,255 @@ func onForward() error {
+// subcommand.
+type configureDatabaseAWSPrintFlags struct {
+	configureDatabaseAWSFlags
+	//
+	policyOnly bool
+	//
```
Unfinished comments.

r0mant commented on tool/teleport/common/teleport.go:
```diff
@@ -490,6 +535,255 @@ func onForward() error {
+typesListLoop:
+	for _, dbType := range f.typesList {
+		for _, supportedDBType := range awsDatabaseTypes {
+			if dbType == supportedDBType {
```
If you invert this to `!=`, you won't need the loop label.

gabrielcorado replied:
We need the outer loop label because the return statement is on it. We still need to `continue` the outer loop, even changing the condition. The only alternative I see here to make things easier to understand is to place in the inner loop in a separated function so that we could have:
```go
for _, dbType := range f.typesList {
	if !isSupported(dbType) {
		return trace.BadParameter("...")
	}
}
```

What do you think?

r0mant replied:
I have a better proposal - we already have `utils.SliceContainsStr` - can we use that instead of implementing double-loop?

r0mant reviewed (commented):
@gabrielcorado Lots of work, nice job! I think all the main pieces are there but I left a few refactoring suggestions which I think will improve the structure and readability of the implementation.

smallinsky commented on lib/configurators/databases/aws.go:
```diff
@@ -0,0 +1,656 @@
+	}
+}
+
+func (a *awsConfigurator) Execute(ctx context.Context) []ExecutionResult {
```
👍  I liked this kind of builder pattern idea.

smallinsky reviewed (commented):
Left some comments. LGTM when remaining comments will be addressed.

gabrielcorado commented:
@r0mant and @smallinsky, thanks for the review; based on your comments, I've made a few changes:
* Moved `UpsertPolicy`, `RetrievePolicy`, `AttachPolicy` and `AttachPolicyBoundary` to the `lib/cloud/aws` package since they were not specific to the configurator;
* Changed the `Configurator` to return a list of `Actions` where each action can be executed (and tested) independently;
* Split the building policy into smaller functions, making it easier to compose and test;

gabrielcorado commented on tool/teleport/common/teleport.go:
```diff
@@ -490,6 +535,255 @@ func onForward() error {
+
+	if flags.boundaryOnly {
+		// Policy boundary is present at the details of the second instruction.
+		fmt.Println(instructions[1].Details)
```
I've added an `IsEmpty` check, which will guarantee those slice items are present.

r0mant commented on lib/configurators/databases/aws.go:
```diff
@@ -0,0 +1,639 @@
+	}
+
+	return []*awslib.Statement{
+		{Effect: awslib.EffectAllow, Actions: []string{"redshift:GetClusterCredentials"}, Resources: resources},
```
Hmm, do we need to include this in the policy? This is what Teleport agent creates itself during auto-IAM configuration, no? And the `resources` loop above probably isn't needed also.

gabrielcorado replied:
You're right. This is similar to the one Teleport generates. We still need to place them on the boundary policy, or Teleport won't have enough permissions to attach to its user/role. And this loop is reducing the number of "wildcard permissions" we could place a default wildcard permission (like the one added when Redshift auto-discovery is enabled).

gabrielcorado replied:
> Hmm does that mean though that they will have to re-run configurator after adding each new database?
For static Redshift database, yes. It is the tradeoff of having a more restrictive boundary.

We can identify that the agent has dynamic registration and go for the wildcard in those cases since we don't know the complete databases list. And since it is the recommended way, going with only wildcard permissions for Redshift sounds ok.

r0mant replied:
Yes, let's do that. Do you need to update this method then to remove the loop above?

r0mant commented on tool/teleport/common/teleport.go:
```diff
@@ -578,6 +625,257 @@ func onForward() error {
+// subcommand.
+type configureDatabaseAWSPrintFlags struct {
+	configureDatabaseAWSFlags
+	//
```
Missing comments here and below.

r0mant reviewed (commented):
Much cleaner now :+1: I have left a few more suggestions.

r0mant reviewed (commented):
@gabrielcorado Also, have you tested this on an actual AWS node with IAM role attached? Let's please test this scenario.

gabrielcorado commented on lib/configurators/databases/aws.go:
```diff
@@ -0,0 +1,639 @@
+		return trace.Wrap(err)
+	}
+
+	formattedPolicy, err := formatPolicyDocument(policyDocument)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	formattedPolicyBoundary, err := formatPolicyDocument(policyBoundaryDocument)
+	if err != nil {
+		return trace.Wrap(err)
+	}
```
It is kept outside so that neither we need an `awsPolicyCreator/awsPoliciesAttacher` constructors nor the `Details` function needs to return an error. For me, it makes sense to keep this close to the policy build since if there is an issue with it, the root cause is "close by".

gabrielcorado commented on lib/configurators/databases/aws.go:
```diff
@@ -0,0 +1,639 @@
+// Execute retrieves policy and policy boundary ARNs from `context.Context` and
+// attach them to the `target`.
+func (a *awsPoliciesAttacher) Execute(ctx context.Context) (context.Context, error) {
+	if a.helper == nil {
+		return ctx, trace.BadParameter("policy helper not initialized")
+	}
```
This check is necessary if the caller tries to execute an Action generated in "manual" mode (meaning it won't have the `Policies` initialized). We could have different structs for when it's manual or not.

gabrielcorado commented on lib/configurators/databases/aws.go:
```diff
@@ -0,0 +1,639 @@
+// attach them to the `target`.
+func (a *awsPoliciesAttacher) Execute(ctx context.Context) (context.Context, error) {
+	if a.helper == nil {
+		return ctx, trace.BadParameter("policy helper not initialized")
```
I've changed this logic to no longer use the `context.Context` as Roman suggested.

r0mant commented on tool/teleport/common/teleport.go:
```diff
@@ -593,6 +641,252 @@ func onCheckHome() error {
+
+// onConfigureDatabaseBootstrap subcommand that bootstraps configuration for
+// database agents.
+func onConfigureDatabaseBootstrap(flags configureDatabaseBootstrapFlags) error {
```
Can we please move all these new methods out of `teleport.go` and into its own source file e.g. `configurator.go`?

r0mant commented on lib/cloud/aws/policy.go:
```diff
@@ -128,11 +158,236 @@ func (p *PolicyDocument) Delete(effect, action, resource string) {
 	p.Statements = statements
 }
 
+// PolicyHelper set of IAM Policy helper functions defined as an interface to
```
```suggestion
// Policies set of IAM Policy helper functions defined as an interface to
```

r0mant reviewed (commented):
A final batch of comments, should be good to go after that.

@gabrielcorado Also, have you seen my comment about trying this out on an actual EC2 node with an IAM role attached? Have you tested this scenario? Let me know if I can help you with it.

r0mant commented on tool/teleport/common/teleport.go:
```diff
@@ -593,6 +602,12 @@ func onCheckHome() error {
 	return nil
 }
 
+// configureDatabaseBootstrapFlags database configure bootstrap flags.
+type configureDatabaseBootstrapFlags struct {
+	config  dbconfigurators.BootstrapFlags
+	confirm bool
+}
```
Should probably move this to `configurator.go` too?

r0mant commented on tool/teleport/common/configurator.go:
```diff
@@ -0,0 +1,307 @@
+		err := action.Execute(ctx, actionContext)
+		printDBBootstrapActionResult(configuratorName, action, err)
+		if err != nil {
+			return err
```
```suggestion
			return trace.Wrap(err)
```

gabrielcorado commented:
@r0mant Yeah, I did the testing with roles attached to EC2 instances to access RDS and Redshift databases.

r0mant committed:
Merge branch 'master' into gabrielcorado/aws-database-configurator
